File: MACRO/app/schemas/acceso.py
"""Esquemas del control de acceso (verificación Oracle + solicitud)."""
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel

EstadoAcceso = Literal[
    "permitido",
    "no_registrado",
    "pendiente",
    "rechazado",
    "inactivo",
    "sin_conexion",
]


class AccesoEstado(BaseModel):
    estado: EstadoAcceso
    usuario_red: str
    mensaje: str


class SolicitudAccesoRequest(BaseModel):
    nombre_completo: str
    # Correo institucional obligatorio; el dominio se valida en el servicio
    # (auth_service.solicitar_acceso) para devolver un mensaje amigable en vez
    # de un 422 de Pydantic.
    email: str


class AccesoResultado(BaseModel):
    ok: bool
    mensaje: str

File: frontend/src/api/acceso.ts
export type EstadoAcceso =
  | "permitido"
  | "no_registrado"
  | "pendiente"
  | "rechazado"
  | "inactivo"
  | "sin_conexion";

export interface AccesoEstado {
  estado: EstadoAcceso;
  usuario_red: string;
  mensaje: string;
}

export interface AccesoResultado {
  ok: boolean;
  mensaje: string;
}

async function pedir<T>(url: string, init?: RequestInit): Promise<T> {
  const resp = await fetch(url, init);
  if (!resp.ok) {
    throw new Error(`Error ${resp.status} en ${url}`);
  }
  return (await resp.json()) as T;
}

export const getAcceso = () => pedir<AccesoEstado>("/api/acceso");

export const solicitarAcceso = (nombre_completo: string, email: string) =>
  pedir<AccesoResultado>("/api/acceso/solicitud", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ nombre_completo, email }),
  });
